[APM] Add a default UDS listeners for traces and stats

ichinaski opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Jira ticket: https://datadoghq.atlassian.net/browse/APMSP-1219

This PR adds default UDS listeners on `/var/run/datadog/apm.socket` (traces, trace-agent) and `/var/run/datadog/dsd.socket` (dogstatsd, core-agent).
 
These default sockets are defined in [this](https://github.com/DataDog/architecture/blob/4cecaf5e0f3d1fbd2c039b37373710fefe8c6465/rfcs/apm/integrations/trace-autodetect-agent-config/rfc.md#overview) RFC, and are used in the Single Step APM Instrumentation and auto-injector. They are meant to minimize config changes and reduce onboarding friction.

If "/var/run/datadog" is not created by systemd (on very old systems), there is no need to create these sockets.

~If the user-configured sockets differs from these default ones, additional UDS listeners will be created in both trace-agent (traces) and core-agent (dogstatsd).~

### Edit July 11:

The existing configuration settings on both APM and statsd have been leveraged for this feature, just updating their default values:
* `apm_condig.receiver_socket` will default to `/var/run/datadog/apm.socket` on unix systems, `""` elsewhere.
* `dogstatsd_socket` will default to `/var/run/datadog/dsd.socket` on unix systems, `""` elsewhere.

This update considerably simplifies the PR and removes the need to have additional listeners, which need to also be controlled on the customer side, resulting in extra complexity in the configuration


<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://datadoghq.atlassian.net/browse/APMSP-1219

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

ichinaski committed:
[APM] Add a default UDS listener for traces

This PR adds a default UDS listener on `/var/run/datadog/apm.socket`.
This is used in the k8s case and SSI on host/docker rely on this socket
path.

If the user-configured socket differs from this default one, 2 different
listeners will be created.

If "/var/run/datadog" is not created by systemd (on very old systems),
there is no need to create this socket.

ichinaski commented on pkg/trace/config/config.go:
```diff
@@ -485,12 +487,14 @@ func New() *AgentConfig {
 		RareSamplerCooldownPeriod: 5 * time.Minute,
 		RareSamplerCardinality:    200,
 
+		ReceiverEnabled:        true,
```
Always true except when manually overridden. The idea behind this flag is to allow OTLP component to disable the entire receiver (for example [here](https://github.com/DataDog/datadog-agent/blob/main/comp/otelcol/otlp/components/statsprocessor/agent.go#L47-L48) )

knusbaum commented on pkg/trace/api/api.go:
```diff
@@ -266,7 +271,27 @@ func (r *HTTPReceiver) Start() {
+	// Listen on the default APM socket
+	if path := r.conf.ReceiverDefaultSocket; path != "" {
+		if _, err := os.Stat(filepath.Dir(path)); !os.IsNotExist(err) {
+			ln, err := r.listenUnix(path)
+			if err != nil {
+				r.telemetryCollector.SendStartupError(telemetry.CantStartUdsServer, err)
+				killProcess("Error creating UDS listener: %v", err)
+			}
+			go func() {
+				defer watchdog.LogOnPanic(r.statsd)
+				if err := r.server.Serve(ln); err != nil && err != http.ErrServerClosed {
+					log.Errorf("Could not start UDS server: %v. UDS receiver disabled.", err)
+					r.telemetryCollector.SendStartupError(telemetry.CantStartUdsServer, err)
+				}
+			}()
+			log.Infof("Listening for traces at unix://%s", path)
```
Is it worth deduplicating this code with the below?

knusbaum reviewed (approved):
LGTM, just one comment.

ichinaski committed:
Small refactor on shared logic, add releasenotes

Extracts common code into a function.

Add releasenotes.

rtrieu reviewed (approved):
Looks good!

ichinaski committed:
Merge branch 'main' into ichinaski/apmsocket

ichinaski committed:
Add default dogstatsd listener

Create a default dogstatsd UDS listener in the expected location
(/var/run/datadog/dsd.socket).

This socket is used by tracers using auto-injection, and is meant to
reduce onboarding friction, minimising the changes needed on the
customer end.

If the customer-defined socket differs from the default one, 2 UDS
listeners are created.

vickenty commented on comp/dogstatsd/server/server.go:
```diff
@@ -373,8 +380,20 @@ func (s *server) start(context.Context) error {
+	if _, err := os.Stat(filepath.Dir(defaultSocket)); !os.IsNotExist(err) {
+		unixListener, err := listeners.NewUDSDatagramListener(packetsChannel, sharedPacketPoolManager, sharedUDSOobPoolManager, s.config, defaultSocket, s.tCapture, s.wmeta, s.pidMap)
+		if err != nil {
+			s.log.Errorf("Can't init listener: %s", err.Error())
```
Would it make sense to specify what kind of listener and which path we tried to create?

vickenty commented on comp/dogstatsd/server/server.go:
```diff
@@ -373,8 +380,20 @@ func (s *server) start(context.Context) error {
-	if len(socketPath) > 0 {
-		unixListener, err := listeners.NewUDSDatagramListener(packetsChannel, sharedPacketPoolManager, sharedUDSOobPoolManager, s.config, s.tCapture, s.wmeta, s.pidMap)
+	// Listen on the default socket (if /var/run/datadog/ exists)
+	if _, err := os.Stat(filepath.Dir(defaultSocket)); !os.IsNotExist(err) {
```
Would it make sense to have an option to disable this in the odd case it causes issues somewhere?

vickenty replied:
Any reason to hide it? I think if there are problems, it would be best if users can discover and enable solution on their own, without hitting support.

ichinaski replied:
I discussed this with the team and the preferred option is to keep the exposed configuration for the client as simple as possible. Unless there is a known issue that opening this socket can cause, we can probably leave it as is. The code already checks for the existence of the parent directory, and should be harmless.

What do you think?

carlosroman replied:
I agree with @vickenty and it would be good to have a way to disable this feature completely as we never know what customers have done to secure their OS or may object to us enabling this on their system by default.

ichinaski replied:
Okay, I think I figured out a good middleground between overloading the config and doing uncontrollable things: Let's use the existing `dogstatsd_socket` and update the default value from `""` (disabled) to `/var/run/datadog/dsd.socket`, but only on non-windows systems. I see that this is a pattern already used in other places of the config template (for instance [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml#L1672-L1684) or [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml#L1841-L1869)).

Does this sound good from statsd perspective?

vickenty commented on comp/dogstatsd/server/server.go:
```diff
@@ -64,6 +66,11 @@ var (
+	// defaultSocket will always be opened as long as the parent directory is available,
+	// usually created by systemd. If the user-configured socket differs from this one,
+	// 2 sockets will be listening.
+	defaultSocket = "/var/run/datadog/dsd.socket"
```
Would it make sense to make this a constant?

ichinaski replied:
I initially had a constant, but switched to a variable to be able to better test it.

vickenty commented on releasenotes/notes/apm-add-default-uds-listener-01abbd2119552312.yaml:
```diff
@@ -0,0 +1,12 @@
+enhancements:
+  - |
+    APM: Add a default UDS listener on /var/run/datadog/apm.socket.
+    This is used in the Single Step APM Instrumentation, improving the overall onboarding experience and minimizing the agent configuration.
```
This should mention changes to dogstatsd too.

ichinaski replied:
good catch!

ichinaski committed:
More detailed error log, update releasenotes

ichinaski committed:
Merge branch 'main' into ichinaski/apmsocket

ichinaski committed:
fix linux test constructor

ichinaski committed:
fix  test constructor

ichinaski committed:
Fix test race conditions

When tests are run in parallel, they were using the same sockets,
potentially causing race conditions and flaky tests.

This fix addresses the issue by creating independent directories for
each test run.

ichinaski committed:
Fix UDS tests

ichinaski committed:
Fix UDS tests

ichinaski committed:
Split UDS tests

ichinaski committed:
Extract UDS tests to standalone file

ichinaski committed:
Merge branch 'main' into ichinaski/apmsocket

ichinaski committed:
Use single setting for APM UDS listeners

Having 2 different UDS listeners in the configuration can be misleading
and adds complexity on the user side.

We will instead keep using a single socket setting, but updating the
default value.

On unix systems, the default values are "/var/run/datadog/apm.socket"
and "/var/run/datadog/dsd.socket"

On other systems, the default value are "" (features disabled).

ajgajg1134 reviewed (approved):
looks good!

GustavoCaso commented on comp/dogstatsd/server/server_test.go:
```diff
@@ -480,6 +468,22 @@ func TestUDPReceive(t *testing.T) {
+	require.NoError(t, err, "cannot connect to UDP network")
+	defer conn.Close()
+
+	testReceive(t, conn, demux)
```
My bad I can see it is used on the other file 😄

GustavoCaso commented on comp/dogstatsd/server/server.go:
```diff
@@ -371,12 +373,14 @@ func (s *server) start(context.Context) error {
-		} else {
-			tmpListeners = append(tmpListeners, unixListener)
-			udsListenerRunning = true
+		if _, err := os.Stat(filepath.Dir(socketPath)); !os.IsNotExist(err) {
```
```suggestion
		if _, err := os.Stat(filepath.Dir(socketPath)); !errors.Is(err, fs.ErrNotExist) {
```

GustavoCaso replied:
Also, should we check if the socket is not closed already? 

```go
if _, err := os.Stat(filepath.Dir(socketPath)); !errors.Is(err, fs.ErrNotExist) && !errors.Is(err, fs. ErrClosed)
```

vickenty replied:
`Stat` on a directory probably won't ever return `ErrClosed`.

@ichinaski Do you think we can remove this entire condition and restore logic the way it was originally? We should always report something if we couldn't create the socket.

GustavoCaso commented on comp/dogstatsd/server/server_uds_test.go:
```diff
@@ -0,0 +1,51 @@
+	testReceive(t, conn, demux)
+}
+
+func TestUDSReceiverDisabled(t *testing.T) {
```
Should we test in the case that socket if socket file exists and it is close?

ichinaski committed:
ensure the socket is closed upon server stop

vickenty commented on comp/dogstatsd/server/server_uds_test.go:
```diff
@@ -0,0 +1,51 @@
+	_ = fulfillDepsWithConfigOverride(t, cfg)
+
+	socketPath := filepath.Join(t.TempDir(), "dsd.socket")
+	_, err := net.Dial("unixgram", socketPath)
```
I may be missing something, this test seems a bit non-sensical. TempDir() creates a new randomly named directory each time it is called, so the name is not known, nor the directory exists when we configure and start the dogstatsd server. Even if the socket is not disabled as expected, it won't be at this random location.

ichinaski replied:
Yeah true. I will just remove this altogether, and rely on the other test that will also stop the server and ensure the conn fails if called after that.

ichinaski replied:
Updated. I actually left the test to check for 
```go
require.False(t, deps.Server.UdsListenerRunning())
```

ichinaski committed:
Use errors.Is instead of os.IsNotExist

Co-authored-by: Gustavo Caso <gustavo.caso@datadoghq.com>

ichinaski committed:
remove useless conn test

ichinaski committed:
add missing imports

ichinaski committed:
fix integration test

vickenty commented on comp/dogstatsd/server/server.go:
```diff
@@ -371,12 +375,14 @@ func (s *server) start(context.Context) error {
-		} else {
-			tmpListeners = append(tmpListeners, unixListener)
-			udsListenerRunning = true
+		if _, err := os.Stat(filepath.Dir(socketPath)); !errors.Is(err, fs.ErrNotExist) {
```
Do you think we can remove this entire condition and restore logic the way it was originally? We should always report something if we couldn't create the socket, both for default and user-provided paths.

ichinaski replied:
Ok, re-reading the code if the parent directory doesn't exist it will just log `"Can't init UDS listener on path %s: %s"`, and continue the server start (no panic). This should be fine regardless of where the value is coming from (default vs explicitly set).

Also, just to clarify, by "restore the logic the way it was originally" you are referring exactly to the point above, not to the original PR where I had a separate listener for the default socket, right?

ichinaski committed:
Always attempt to create the UDS listener

Do not silently skip the UDS listener if the parent dir doesn't exist.

Instead, we will revert this extra check, and log an error if the socket
cannot be created (as it was before).

vickenty commented on comp/dogstatsd/server/server_uds_test.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+func TestUDSReceiverNoDir(t *testing.T) {
+	socketPath := "/var/run/datad0g/dsd.socket" // nonexistent dir, listener should not be set
```
```suggestion
	socketPath := filepath.Join(t.TempDir(), "nonexistent", "dsd.socket") // nonexistent dir, listener should not be set
```

ichinaski committed:
Update socket path in test case

Co-authored-by: Vickenty Fesunov <vickenty@users.noreply.github.com>

ichinaski committed:
add error log if dir is missing

GustavoCaso reviewed (approved):
LGTM

ichinaski commented:
/merge
